modules/internal/ChapelArray.chpl

```diff
@@ -2939,13 +2991,189 @@ module ChapelArray {
+  // helpers, but seemingly, something about returning an array view
+  // from a procedure that calls another procedure to create it didn't
+  // work (?).  Though I also wasn't easily able to reproduce this in
+  // a standalone test, so not sure what's going on there...
+  //
+  pragma "no promotion when by ref"
+  pragma "reference to const when const this"
+  proc _array.chpl_copyReshape(dom: domain(?),
+                               checkDims=checkReshapeDimsByDefault) {
+    if !dom.isRectangular() then
+      compilerError("reshape() with copying is currently only supported for rectangular domains");
+    if boundsChecking then
+      chpl__validateReshape(this, dom);
+    // TODO: Add a parallel linearize() iterator to all rectangular types
+    // and zip those instead of using this serial implementation
+    var B: [dom] this.eltType = for (i,a) in zip(dom, this) do a;
+    return B;
+  }
+
+  pragma "no promotion when by ref"
+  pragma "reference to const when const this"
+  pragma "fn returns aliasing array"
+  proc _array.chpl_aliasReshape(dom: domain(?),
+                                checkDims=checkReshapeDimsByDefault) {
+    if chpl__isArrayView(this) ||
+       !Reflection.canResolveMethod(this._value, "doiSupportsReshape") {
+         compilerError("This array type does not support alias-based reshaping; consider passing 'copy=true' to get a copy-based reshape");
+       return this;
+    } else if !Reflection.canResolveMethod(this._value, "doiReshape", dom) {
+      compilerError("This array cannot be reshaped using this domain type; consider passing 'copy=true' to get a copy-based reshape");
+      return this;
+    } else {
+      if boundsChecking && checkDims then
+        chpl__validateReshape(this, dom);
+      return this._value.doiReshape(dom);
+    }
+  }
+
+  proc chpl__validateReshape(arr, dom) {
+    if dom.size != arr.size then
+      halt("Size mismatch: Can't rehape a ", arr.size,
+           "-element array into a ", dom.size, "-element array");
+
+    if arr.size > 0 && dom.size > 0 {
+      var arrDim, domDim = 0;
+      var arrSize, domSize = 0;
+      var arrCatchup, domCatchup, error = false;
+      do {
+        if arrSize == 0 {
+          arrSize = arr.dim(arrDim).size;
+        }
+        if domSize == 0 {
+          domSize = dom.dim(domDim).size;
+        }
+        if (arrSize == domSize) {
+          arrDim += 1;
+          domDim += 1;
+          arrSize = 0;
+          domSize = 0;
+          arrCatchup = false;
+          domCatchup = false;
+        } else if (arrSize < domSize) {
+          if domCatchup then
+            error = true;
+          arrCatchup = true;
+          arrDim += 1;
+          arrSize *= arr.dim(arrDim).size;
+        } else {
+          if arrCatchup then
+            error = true;
+          domCatchup = true;
+          domDim += 1;
+          domSize *= dom.dim(domDim).size;
+        }
+      } while (arrDim < arr.rank && domDim < dom.rank);
```

Comment:
nit: parens (I actually can't remember if we need them after while do but I'd assume not? considering our normal stance on parens 😄)

Reply:
You're correct that we don't.